### PR TITLE
git tag trigger for pipelines

### PR DESCRIPTION
**Pull request by gitlawr**
Add support for triggering pipelines by git tag
Related issue: https://github.com/rancher/rancher/issues/12624
Dependent types PR: https://github.com/rancher/types/pull/376

### CODE REVIEW

File: pkg/controllers/user/pipeline/remote/github/github.go [hunk 1]
```diff
@@ -104,13 +104,7 @@ func (c *client) CreateHook(pipeline *v3.Pipeline, accessToken string) (string,
...
 		return "", err
 	}
-	events := []string{}
-	if pipeline.Spec.TriggerWebhookPush {
-		events = append(events, "push")
-	}
-	if pipeline.Spec.TriggerWebhookPr {
-		events = append(events, "pull_request")
-	}
+	events := []string{"push", "pull_request"}
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/remote/github/github.go]: why we create events for push and pull_request?

File: pkg/controllers/user/pipeline/hooks/drivers/github.go [hunk 2]
```diff
@@ -69,14 +69,29 @@ func (g GithubDriver) Execute(req *http.Request) (int, error) {
 	}
 
 	ref, branch, commit := "", "", ""
+	optionalEnvVars := map[string]string{}
```
Comment (zionwu) [inline pkg/controllers/user/pipeline/hooks/drivers/github.go]: Why we need optionalEnvVars to store tag? what is the different between tag and commit?

File: pkg/controllers/user/pipeline/remote/github/github.go (hunk 1 above)
Comment (gitlawr) [reply pkg/controllers/user/pipeline/remote/github/github.go]: We create github webhook with all needed events.
When it is triggered by any of these events. Hook handler check the pipeline config and choose to run it or ignore it.
In this way, we don't need to carefully inspect and update webhook config in github repo everytime the pipeline is updated(when some of the triggers can be enabled/disabled)

File: pkg/controllers/user/pipeline/hooks/drivers/github.go (hunk 2 above)
Comment (gitlawr) [reply pkg/controllers/user/pipeline/hooks/drivers/github.go]: Tag is not included in pipeline execution fields while commit is.

**Comment by zionwu** [comment]
LGTM

File: pkg/controllers/user/pipeline/hooks/drivers/github.go [hunk 3]
```diff
@@ -69,14 +69,29 @@ func (g GithubDriver) Execute(req *http.Request) (int, error) {
...
 	if event == "push" {
 		payload := &github.PushEvent{}
 		if err := json.Unmarshal(body, payload); err != nil {
 			return http.StatusUnprocessableEntity, err
 		}
 		ref = payload.GetRef()
-		branch = strings.TrimLeft(payload.GetRef(), "refs/heads/")
-		commit = payload.HeadCommit.GetID()
+		if strings.HasPrefix(ref, "refs/tags/") {
```
Comment (alena1108) [inline pkg/controllers/user/pipeline/hooks/drivers/github.go]: can we use const for the prefix? since it repeats in a couple of places

**Pull request merged**